### PR TITLE
Release/0.3.1

### PR DESCRIPTION
**Pull request by asfaltboy**
Hi there,

This pull request should fix the lecture fetching mechanism issue #4 , to update with recent changes.

While it is mostly based on your master, I've made a couple of major changes to the page and cookie fetching mechanism, using the `requests` package (instead of using `urrlib` directly). Additionally I added a LICENSE file, a changelog and refactored most of the code to PEP8 spec, aiming to make it simple and readable.
However, the majority of the syllabus parsing, and the actual `addon` functionality was left as is.

If you like, you could merge the pull request to your master, or tell me what you would like me to amend so it more closely matches what you're looking for. For me though, this works well enough.

Regards,
Pavel

### CODE REVIEW

File: addon.xml [hunk 1]
```diff
@@ -1,9 +1,10 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<addon id="plugin.video.coursera" name="Coursera" version="0.302" provider-name="Alex Rothberg (cancan101)">
+<addon id="plugin.video.coursera" name="Coursera" version="0.3.1" provider-name="Alex Rothberg (cancan101)">
   <requires>
     <import addon="xbmc.python" version="2.0"/>
-    <import addon="script.module.xbmcswift2" version="1.1.1"/>
-    <import addon="script.module.beautifulsoup" version="3.0.8"/>
+    <import addon="script.module.xbmcswift2" version="2.4.0"/>
```
Comment (cancan101) [inline addon.xml]: Where did you see that `v2.4.0` is right for xbmcswift2?

**Comment by cancan101** [comment]
I would prefer a more permissive license like BSD or MIT.

File: coursera_login.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    user_data = plugin.get_storage(username, file_format='json')
+
+    cookies_class = user_data.get('cookies_class')
+    if cookies_class is None:
+        cookies_class = user_data['cookies_class'] = {}
+
+    return cookies_class
+
+
+def getClassCookieOrLogin(username, password, courseShortName,
```
Comment (cancan101) [inline coursera_login.py]: You can probably drop the silliness of `indicateDidLogin`. That was laziness on my part but it looks like your refactor might make it unneeded.

File: coursera_login.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    cookies_raw = user_data.get('cookies')
+    if cookies_raw is None:
+        cookies_raw = login(username, password)
+        if cookies_raw is None:
+            return None
+        user_data['cookies'] = cookies_raw
+
+    res = requests.get(get_auth_url(className), allow_redirects=False,
+                       cookies=cookies_raw)
+    if not res.ok:
```
Comment (cancan101) [inline coursera_login.py]: The `res.ok` check is redundant (see: https://github.com/kennethreitz/requests/blob/bd3cf95e34aa49c8d764c899672048df107e0d70/requests/models.py#L617). Just call `res.raise_for_status()`.

File: course_utils.py [hunk 4]
```diff
@@ -3,204 +3,352 @@
...
+                if resource_format == 'mp4':
+                    mp4_found = True
+                    resources_entry["Lecture Video"] = href
+                elif resource_format == 'srt':
+                    resources_entry["Subtitle"] = href
+
+                resources_entry[title] = href
+
+            if not mp4_found:
+                plugin.log.error("No MP4 resource found. Using hidden video "
```
Comment (cancan101) [inline course_utils.py]: What hidden video logic? That seems to longer be called.

**Comment by cancan101** [comment]
Feel free to add a contributors file and list yourself in it.

File: course_utils.py [hunk 5]
```diff
@@ -3,204 +3,352 @@
...
+        }, ...  # more sessions ],
+    }, ...  # more classes }
+    """
+    classes = {}
+    course_metadata = {}
+    for course in classes_data['linked']['courses.v1']:
+        course_metadata[course['id']] = course
+
+    sorted_sessions = sorted(classes_data['linked']['v1Sessions.v1'],
+                             key=operator.itemgetter('id'), reverse=True)
```
Comment (cancan101) [inline course_utils.py]: why not just `lambda x: x['id']`?

File: course_utils.py [hunk 6]
```diff
@@ -3,204 +3,352 @@
...
+        user_data['cookies'] = cookies_raw
+
+        classes = getClasses(cookies=cookies_raw)
+
+    return classes
+
+
+def get_page(href, json=False, **kwargs):
+    res = requests.get(href, **kwargs)
+    if not res.ok:
```
Comment (cancan101) [inline course_utils.py]: ditto

**Comment by cancan101** [comment]
Otherwise, looks good.

File: addon.xml (hunk 1 above)
Comment (asfaltboy) [inline addon.xml]: [over here](http://kodi.wiki/view/Add-on:Xbmcswift2) and in xbmc itself after installing xbmcswift2 manually (in settings -> add-ons -> add-on libraries)

**Comment by asfaltboy** [comment]
Thank you for the review, I'll try and commit a fix with your changes later today

File: course_utils.py (hunk 5 above)
Comment (asfaltboy) [inline course_utils.py]: It's basically the same; just got used to using operator functions

File: LICENSE.txt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) [year] [fullname]
```
Comment (cancan101) [inline LICENSE.txt]: Should fill in the year and name, I think.

Comment (asfaltboy) [inline LICENSE.txt]: Didn't want to obtrude here, is 2015 and Alex Rothberg good enough?

Comment (cancan101) [inline LICENSE.txt]: Sounds good

**Comment by medyagh** [comment]
Hey I tried this new PR, it still has the the dependency problem

**Comment by cancan101** [comment]
@medyagh Can you past w/e is in xbmc.log.

**Comment by medyagh** [comment]
@cancan101
```
  21:01:27 T:140212976765120   DEBUG: ProcessMouse: trying mouse action leftclick
21:01:27 T:140212976765120   DEBUG: ------ Window Deinit (Pointer.xml) ------
21:01:27 T:140212976765120   DEBUG: ------ Window Init (Pointer.xml) ------
21:01:28 T:140212976765120   DEBUG: ProcessMouse: trying mouse action leftclick
21:01:28 T:140212976765120   DEBUG: ------ Window Deinit (Pointer.xml) ------
21:01:28 T:140212976765120   DEBUG: ------ Window Deinit (FileBrowser.xml) ------
21:01:28 T:140212976765120   DEBUG: Addon plugin.video.coursera requires xbmc.python version 2.0 which is not available
21:01:28 T:140212976765120   DEBUG: ------ Window Init (DialogKaiToast.xml) ------
21:01:29 T:140212976765120   DEBUG: Keyboard: scancode: 0x40, sym: 0x0134, unicode: 0x0000, modifier: 0x100
21:01:29 T:140212976765120   DEBUG: OnKey: alt-leftalt (0x4f0d4) pressed, action is
21:01:34 T:140212976765120   DEBUG: ------ Window Deinit (DialogKaiToast.xml) ------
```

**Comment by cancan101** [comment]
Are you sure that you are using the code in the PR? The Python version was changed to 2.1.0: https://github.com/cancan101/xbmc-coursera/pull/6/files#diff-a6d9d350408378c9065d9fb893cc1e69L4

**Comment by medyagh** [comment]
@cancan101  oops you were right I was uploading older version for some reason!  I confirm this PR works just fine, sorry about the mistake.

**Comment by medyagh** [comment]
@cancan101  can I make an icon for this github it ?

**Comment by cancan101** [comment]
Sure, just throw up a new PR.

**Comment by asfaltboy** [comment]
Sorry for the delay guys, I pushed a few more updates (mostly the license fix and clearer feature description).

@medyagh out of curiousity, which version of xbmc are you using?

The [addon documentation](http://kodi.wiki/view/Addon.xml#.3Crequires.3E) state that the version in the requirements should be the minimum version required by the addon.

So it should have worked either way for any version >= 2.0 (given that 2.1 works for you and 2.0 < 2.1). And indeed it works for me if I try any value < 2.15.0, being that I'm on Kodi v14.1

**Comment by cancan101** [comment]
LMK if you plan to do any more work, otherwise I will merge.

**Comment by asfaltboy** [comment]
Nothing else planned for now, thanks

**Comment by cancan101** [comment]
@asfaltboy Thanks!

**Comment by DmitrySandalov** [comment]
Brilliant, works for me! Thanks for PEP8 improvements.

**Comment by asfaltboy** [comment]
Thanks everyone for your help! Glad it works for you @DmitrySandalov .

Future TODO list for work on the addon we should consider (in my opinion) is:
  1. check out issue #10 and #4
  2. complete work on icon feature #8
  3. add folders/filters for course: current / previous / unattended, etc
  4. implement "join course" from within addon (accept course agreement)
  5. prepare for and submit plugin to official Kodi repo (for others to easily find it)

I'll try to find time for some, but please do submit a PR if you happen to do it earlier.

**Pull request merged**